src/lib/Microsoft.Health.Events/EventConsumers/Service/EventBatchingService.cs

```diff
@@ -148,11 +154,21 @@ private async Task CompleteProcessing(IEnumerable<IEventMessage> events)
             }
 
             await UpdateCheckpoint(events, eventMetrics);
+
+            LogDataFreshness(partitionId, triggerReason, events);
         }
 
         public Task ConsumeEvents(IEnumerable<IEventMessage> events)
         {
             throw new NotImplementedException();
         }
+
+        private void LogDataFreshness(string partitionId, string triggerReason, IEnumerable<IEventMessage> events = null)
+        {
+            // To determine the data freshness per partition (i.e. latest event data processed in a partition), use the enqueued time of the last event for the batch.
+            // If no events were flushed for the partition (eg: trigger reason is ThresholdWaitReached - due to receival of MaxTimeEvent), then use the current timestamp.
+            var eventTimestampLastProcessed = events?.Any() ?? false ? events.Last().EnqueuedTime.UtcDateTime : DateTime.UtcNow;
+            _logger.LogMetric(EventMetrics.EventTimestampLastProcessedPerPartition(partitionId, triggerReason), long.Parse(eventTimestampLastProcessed.ToString("yyyyMMddHHmmss")));
```

Comment:
We can use double for this. I thought to change the method signature in IomtTelemetryLogger since I noticed that the underlying TelemetryClient has overloads for object when saving the metric so thought we can expose to more types too, but if there are limitations in MDM we want to honor then I'll revert back to double.